Change HandleUntyped::typed to not panic

kroltan opened the pull request:
# Objective

Fixes #2534

## Solution

Changed the signature and implementation of `HandleUntyped::typed` to return a `Result` instead of panicking.

Adjusted a bunch of engine crate code that uses that function, but I sense a code smell. All usages are creating `HandleUntyped`s statically, and converting to `Handle<T>` on use. This could be avoided by just creating a `Handle<T>` in the first place.

I'll open another PR with that change.

alice-i-cecile commented:
bors try

alice-i-cecile commented:
> Adjusted a bunch of engine crate code that uses that function, but I sense a code smell. All usages are creating HandleUntypeds statically, and converting to Handle<T> on use. This could be avoided by just creating a Handle<T> in the first place.

@cart do you have insight on this? Will the new renderer changes alter this?

kroltan commented:
Hmm seems I done a silly one and forgot to run `cargo fmt`.

kroltan commented:
Actually I forgot to run the whole CI. Still some elbow grease to go on this one!

kroltan commented:
@alice-i-cecile With this change, [`AssetServer::load`](https://github.com/bevyengine/bevy/pull/2536/files#diff-000c56acac9637f360f01718f84ef7384488e0558c3e42ab781e760305a2ba58R232) changes to propagate the `Result`.

However this raises a further question, how come it wasn't a `Result` before? It seems that internally it goes to `load_untracked` that just yells a warning if the loading fails.

This feels like a big API change, so should I do the same and output a warning instead of changing the return type of `AssetServer::load`?

alice-i-cecile commented:
bors try

alice-i-cecile commented:
> However this raises a further question, how come it wasn't a Result before? It seems that internally it goes to load_untracked that just yells a warning if the loading fails.

Hmm. I'm genuinely unsure. I know asset loading is deliberately async; maybe it's to avoid propagating results?

In any case, I think we should split this work out and create a new issue. The asset pipeline needs some serious design work in general (see #1701, #708) so it's better to tackle that more holistically.

kroltan commented:
Actually, upon a bit more careful investigation, currently the error path would never trigger anyways. `typed` can only fail for `HandleId::Id`, not `HandleId::AssetPathId`, which is what the `AssetServer` code path uses.

Anyways, fixed the CI errors and reverted the `AssetServer::load` signature.

mockersf commented:
> This could be avoided by just creating a `Handle<T>` in the first place.

There isn't yet a const way to make a `Handle<T>`, but you can probably adapt https://github.com/bevyengine/bevy/blob/a4e5e2790ea00acfcb53ff0d9bdfb09bc93e445c/crates/bevy_asset/src/handle.rs#L244

kroltan commented:
> There isn't yet a const way to make a `Handle<T>`

Yeah, you're right, I [scratched up this change](https://github.com/kroltan/bevy/commit/aa9a52f5eece0f5795ffbb21d91c1215b543a686) but it requires `#![feature(const_fn_trait_bound)]`.

It could be revisited once that language feature get stabilized.

mockersf commented:
> Regarding the contents of this PR, this should be ready for review.

Not a fan of the `FIXME` comments, could you reword them?

kroltan commented:
How about this?

> Handle is always converted to typed, see discussion in PR #2536

mockersf commented:
Good for me! Thanks

NathanSWard commented:
Curious, I'm pretty sure handle used to return a `Result`/`Option` type and then it got changed to use `expect`. 
Do we know the reason for the inital change?

NathanSWard commented:
Could we add unit tests to bind the `HandleUntyped::typed` behavior?

alice-i-cecile commented:
> Curious, I'm pretty sure handle used to return a Result/Option type and then it got changed to use expect.
Do we know the reason for the inital change?

Poking around, it looks like this got changed for 0.3, from https://docs.rs/bevy/0.2.0/bevy/prelude/struct.Handle.html to https://docs.rs/bevy/0.3.0/bevy/asset/struct.HandleUntyped.html#method.typed. Here's the line / commit to blame: https://github.com/bevyengine/bevy/commit/c32e637384936d973918fb59d1c37c0a4f4f41f7#diff-cffa1459c07909ff31710f4d23192ae34a30938951b42076f162702f04c3f601L79

The [PR description](https://github.com/bevyengine/bevy/pull/693) has no justification of this change; @cart did you have some Extremely Good Argument against making this return a Result/Option that we've overlooked?

cart commented:
No "extremely good argument". Just the argument that the implicit panic is more ergonomic. I'm down to change it, given how uncommon this scenario is.

alice-i-cecile commented:
Closing this out, as I don't think it's relevant any more after https://github.com/bevyengine/bevy/pull/4794 :)
